### PR TITLE
No cost timer

### PR DESCRIPTION
**Pull request by mkstoyanov**
This relates to #521 

Switching many of the calls to timer start/stop with raii style of structs (start is called by the constructor and stop is called by the destructor). The CMake option (comes later) will replace the struct with a no-opt one.

### CODE REVIEW

File: src/tools.hpp [hunk 1]
```diff
@@ -106,4 +106,28 @@ class simple_timer
 
 extern simple_timer timer;
 
+/*!
+ * Allows for RAII style of timing for blocks of code.
+ * The constructor will initiate the timer for the given event,
+ * the destructor will stop the timer.
+ */
+struct time_event
```
Comment (quantumsteve) [inline src/tools.hpp:114]: Would you add a unit test in `tools_tests.cpp`?

File: src/tools.hpp [hunk 2]
```diff
@@ -106,4 +106,28 @@ class simple_timer
...
+{
+  //! \brief Constructor, start timing.
+  time_event(std::string event_name)
```
Comment (quantumsteve) [inline src/tools.hpp]:
```suggestion
  time_event(std::string const &event_name)
```

File: src/tools.hpp [hunk 3]
```diff
@@ -106,4 +106,28 @@ class simple_timer
...
+      : event_name_(timer.start(event_name)), flops(-1)
+  {}
+  //! \brief Constructor, start timing for flop count.
+  time_event(std::string event_name, int op_flops)
```
Comment (quantumsteve) [inline src/tools.hpp]:
```suggestion
  time_event(std::string const &event_name, int op_flops)
```

File: src/tools.hpp (hunk 2 above)
Comment (mkstoyanov) [reply src/tools.hpp]: The `event_name` is usually short string and I want to use it inline and move out of it.

Isn't passing by non-const value better for that purpose?

File: src/time_advance.cpp [hunk 4]
```diff
@@ -212,11 +212,14 @@ explicit_advance(PDE<P> const &pde, matrix_list<P> &operator_matrices,
...
-  auto const apply_id = tools::timer.start("kronmult");
   fk::vector<P> fx(row_size);
-  operator_matrices[matrix_entry::regular].apply(1.0, x.data(), 0.0, fx.data());
 
-  tools::timer.stop(apply_id, operator_matrices[matrix_entry::regular].flops());
+  {
+    tools::time_event performance("kronmult");
+    operator_matrices[matrix_entry::regular].apply(1.0, x.data(), 0.0,
+                                                   fx.data());
+    performance.flops = operator_matrices[matrix_entry::regular].flops();
```
Comment (quantumsteve) [inline src/time_advance.cpp:221]: Copying flops like this looks error-prone. (don't need to change in this PR).

File: src/tools.hpp (hunk 2 above)
Comment (quantumsteve) [reply src/tools.hpp]: Assuming SSO and time_event's implementation in the header, I think passing by value or reference will give very similar runtime. In the worst case scenario (size() greater than fits inside `std::string`) passing by const reference would be preferred.

Another alternative would be `std::string_view`, but that would require an overload to `simple_timer::start(...)` and changing the comparator in `id_to_start_` to allow heterogeneous comparison lookup.

File: src/time_advance.cpp (hunk 4 above)
Comment (mkstoyanov) [reply src/time_advance.cpp:221]: You're right, I missed that.

File: src/tools.hpp (hunk 1 above)
Comment (mkstoyanov) [reply src/tools.hpp:114]: Just did.

**Review by quantumsteve — DISMISSED** [review]

File: src/tools.hpp (hunk 2 above)
Comment (mkstoyanov) [reply src/tools.hpp]: You're right, the implementation of the `simple_timer` makes the point mute.

I'm not changing `simple_timer` in this PR.

File: src/tools_tests.cpp [hunk 5]
```diff
@@ -25,20 +25,37 @@ double shuffle_random(int const num_items)
...
   int const items_to_gen       = 100000;
   int const iterations         = 10;
   std::string const identifier = "waste_time";
   for (int i = 0; i < iterations; ++i)
   {
-    timer.start(identifier);
-    double const val = shuffle_random(items_to_gen);
-    timer.stop(identifier);
-    expect(val > 0.0); // to avoid comp. warnings
+    if (std::is_same_v<TestType, simple_timer_tag>)
```
Comment (quantumsteve) [inline src/tools_tests.cpp]:
```suggestion
    if constexpr (std::is_same_v<TestType, simple_timer_tag>)
```

File: src/tools.hpp [hunk 6]
```diff
@@ -106,4 +106,28 @@ class simple_timer
...
+ */
+struct time_event
+{
+  //! \brief Constructor, start timing.
+  time_event(std::string const &event_name)
+      : event_name_(timer.start(event_name)), flops(-1)
+  {}
+  //! \brief Constructor, start timing for flop count.
+  time_event(std::string const &event_name, int op_flops)
+      : event_name_(timer.start(event_name)), flops(op_flops)
```
Comment (quantumsteve) [inline src/tools.hpp:122]: Is the cast from `int` to `double` intentional?

**Review by quantumsteve — CHANGES_REQUESTED** [review]
Just a comment about making if statement (explicitly) constexpr and asking about a cast between types.

File: src/tools.hpp (hunk 6 above)
Comment (mkstoyanov) [reply src/tools.hpp:122]: No, it wasn't. Good catch.

**Review by quantumsteve — APPROVED** [review]

**Pull request merged**